bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,550 @@
+import org.openhab.binding.adorne.internal.AdorneDeviceState;
+import org.openhab.binding.adorne.internal.configuration.AdorneHubConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+
+/**
+ * The {@link AdorneHubController} manages the interaction with the Adorne hub. The controller maintains a connection
+ * with the Adorne Hub and listens to device changes and issues device commands. Interaction with the hub is performed
+ * asynchronously through REST messages.
+ *
+ * @author Mark Theiding - Initial Contribution
+ */
+@NonNullByDefault
+public class AdorneHubController implements Runnable {
+    private Logger logger = LoggerFactory.getLogger(AdorneHubController.class);
+
+    private static final int HUB_CONNECT_TIMEOUT = 15000;
+    private static final int HUB_RECONNECT_SLEEP_MINIMUM = 1;
+    private static final int HUB_RECONNECT_SLEEP_MAXIMUM = 15 * 60;
+
+    // Hub rest commands
+    private static final String HUB_REST_SET_ONOFF = "{\"ID\":%d,\"Service\":\"SetZoneProperties\",\"ZID\":%d,\"PropertyList\":{\"Power\":%b}}\0";
+    private static final String HUB_REST_SET_BRIGHTNESS = "{\"ID\":%d,\"Service\":\"SetZoneProperties\",\"ZID\":%d,\"PropertyList\":{\"PowerLevel\":%d}}\0";
+    private static final String HUB_REST_REQUEST_STATE = "{\"ID\":%d,\"Service\":\"ReportZoneProperties\",\"ZID\":%d}\0";
+    private static final String HUB_REST_REQUEST_ZONES = "{\"ID\":%d,\"Service\":\"ListZones\"}\0";
+    private static final String HUB_REST_REQUEST_MACADDRESS = "{\"ID\":%d,\"Service\":\"SystemInfo\"}\0";
+    private static final String HUB_TOKEN_SERVICE = "Service";
+    private static final String HUB_TOKEN_ZID = "ZID";
+    private static final String HUB_TOKEN_PROPERTY_LIST = "PropertyList";
+    private static final String HUB_TOKEN_DEVICE_TYPE = "DeviceType";
+    private static final String HUB_TOKEN_SWITCH = "Switch";
+    private static final String HUB_TOKEN_DIMMER = "Dimmer";
+    private static final String HUB_TOKEN_NAME = "Name";
+    private static final String HUB_TOKEN_POWER = "Power";
+    private static final String HUB_TOKEN_POWER_LEVEL = "PowerLevel";
+    private static final String HUB_TOKEN_MAC_ADDRESS = "MACAddress";
+    private static final String HUB_TOKEN_ZONE_LIST = "ZoneList";
+    private static final String HUB_SERVICE_REPORT_ZONE_PROPERTIES = "ReportZoneProperties";
+    private static final String HUB_SERVICE_ZONE_PROPERTIES_CHANGED = "ZonePropertiesChanged";
+    private static final String HUB_SERVICE_LIST_ZONE = "ListZones";
+    private static final String HUB_SERVICE_SYSTEM_INFO = "SystemInfo";
+
+    private @Nullable Future<?> hubController;
+    private String hubHost;
+    private int hubPort;
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable Scanner hubIn;
```

Comment:
Don't use `Scanner` everything you read through it can't be garbage collected. I found that out the hard way.
I fiddled around and found that JsonStreamParser is actually a suitable replacement.

Reply:
I had no idea. Made the switch. Had to swap all the downtream Json handling/parsing accordingly - is also a bit cleaner as it's no longer just Map/String based.